cadence/contracts/AxelarGateway.cdc

```diff
@@ -0,0 +1,312 @@
+            self.executedCommandIds.append(commandId)
+            AxelarAuthWeighted.transferOperatorship(params: transferOperatorShipParams!)
+            emit Executed(commandId: commandId)
+          }
+        }
+      } else if (commandHash == self.SELECTOR_APPROVE_CONTRACT_CALL && params[i].length == 6) {
+        let approveContractCallParams = paramHandler.generateApproveContractCallParams()
+
+        if approveContractCallParams != nil {
+          // store the CGPCommand Resource for later use
+          let oldCgpCommand <- self.approvedCommands[commandId] <- create CGPCommand(commandId: commandId)
+
+          emit ContractCallApproved (
+            commandId: commandId,
+            sourceChain: approveContractCallParams!.sourceChain,
+            sourceAddress: approveContractCallParams!.sourceAddress,
+            contractAddress: approveContractCallParams!.contractAddress,
+            payloadHash: approveContractCallParams!.payloadHash,
+            sourceTxHash: approveContractCallParams!.sourceTxHash,
+            sourceEventIndex: approveContractCallParams!.sourceEventIndex
+          )
+
+          destroy oldCgpCommand
+        }
+      } else {
+        continue
+      }
+
+      i = i + 1
+    }
+  }
+
+  access(all) fun executeApp(commandId: String, sourceChain: String, sourceAddress: String, contractAddress: String, payload: [UInt8]): Bool {
+    // Check to see if command id has already been executed
+    if (self.executedCommandIds.contains(commandId)) {
+      return false
+    }
+
+    // Validate that contract address string is a valid address
+    let address = Address.fromString(contractAddress) ?? panic("Invalid address")
+
+    // Verify that the executable capability exists for that contract address
+    let appCapability: &Capability<&{AxelarGateway.Executable}>? = self.account.borrow<&Capability<&{AxelarGateway.Executable}>>(from: self.getAppCapabilityPathFromAddress(address)) ?? self.claimAppCapability(provider: address)
+    if appCapability == nil || !appCapability!.check() {
+      panic("Cannot retrieve app executable capability")
+    }
+
+    // Ensure that the command id is approved
+    if self.approvedCommands[commandId] == nil {
+      panic("Command with id: ".concat(commandId).concat(" does not exist"))
+    }
+
+    // Get a reference to the CGPCommand resource to pass into the executable method
+    let cgpCommand = (&self.approvedCommands[commandId] as &CGPCommand?) ?? panic("Could not borrow reference to CGP Command")
+
+    // Ensure that this cgp command has not been executed before
+    var isExecuted = cgpCommand.isExecuted
+    if !isExecuted {
+      let ref = appCapability!.borrow()!.executeApp(commandResource: cgpCommand, sourceChain: sourceChain, sourceAddress: sourceAddress, payload: payload)
+
+      // Remove the cgp command resource and destroy the resource
+      destroy <- self.approvedCommands.remove(key: commandId)
```

Comment:
See above comment in `CGPCommand.destroy()` about a dedicated contract method to destroy these resources & emit an event